dnf5-plugins/config-manager_plugin/addrepo.cpp

```diff
@@ -0,0 +1,462 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "addrepo.hpp"
+
+#include "shared.hpp"
+
+#include <curl/curl.h>
+#include <fmt/format.h>
+#include <libdnf5/conf/const.hpp>
+#include <libdnf5/repo/file_downloader.hpp>
+#include <libdnf5/repo/repo_query.hpp>
+#include <libdnf5/utils/bgettext/bgettext-mark-domain.h>
+
+#include <filesystem>
+#include <regex>
+#include <string_view>
+
+namespace dnf5 {
+
+using namespace libdnf5::cli;
+
+namespace {
+
+const std::filesystem::path CFG_MANAGER_REPOS_OVERRIDE_FILENAME = "99-config_manager.repo";
+
+const std::filesystem::path CFG_MANAGER_REPOS_OVERRIDE_FILEPATH =
+    libdnf5::REPOS_OVERRIDE_DIR / CFG_MANAGER_REPOS_OVERRIDE_FILENAME;
+
+
+// Extracts a specific part of the URL from a URL string.
+std::string get_url_part(const std::string & url, CURLUPart what_part) {
+    std::string ret;
+    CURLUcode rc;
+    CURLU * c_url = curl_url();
+    rc = curl_url_set(c_url, CURLUPART_URL, url.c_str(), 0);
+    if (!rc) {
+        char * part;
+        rc = curl_url_get(c_url, what_part, &part, 0);
+        if (!rc) {
+            ret = part;
+            curl_free(part);
+        }
+    }
+    curl_url_cleanup(c_url);
+    return ret;
+}
+
+
+// Computes CRC32 checksum of input string.
+// Slow bitwise implementation. Used to calculate the checksum of the omitted part of long URLs.
+uint32_t crc32(std::string_view input) {
```

Comment:
Just a question - would be possible to replace that code by something from a standard library?

Reply:
If I understand correctly, this is used in a specific use case only for generating shorter URLs. Why not utilize a hash function (like SHA1, etc.) from the existing dependency `libcrypto` and extract a substring of the appropriate length from the result?

Reply:
@jan-kolarik I don't currently know why, but there is indeed a dependency on `libcrypto` in "dnf5.spec". There is a line: `BuildRequires: pkgconfig(libcrypto)`.
Are we really using `libcrypto` in libdnf5? Yes `libcurl` which we depend on is compiled against `libcrypto` in Fedora. But libcurl also supports other crypto libraries.

Reply:
I was curious, so I've queued [a testing scratch build](https://koji.fedoraproject.org/koji/taskinfo?taskID=108418756) without `BuildRequires: pkgconfig(libcrypto)` dependency and it really seems unused.

Reply:
Please don't use SHA1 for any purpose. It is not available in FIPS mode in RHEL9.

Reply:
I am not sure whether it is a good idea but what about to use SHA2 but from that long hash use only few first characters. I believe that it is available from libsolv API.

Reply:
Well, SHA-256 is also not recommended by NIST now. In general, a choice of hashing algorithm depends on what you want to use the hash for.

If it is for cryptographic purposes, then it you should go with something implemented by a certified library and be prepared to change the algorithm during a life cycle of the code (i.e. always handle hash value and hash algorithm as a pair in your data and code).

If it is for mapping data onto memory buckets, e.g. for implementing associative arrays, then choose an algorithm witch is designed for it (i.e. fair and fast). E.g. MurmurHash. MD5 and SHA-1 are needlessly slow.

If it is for error detection, then CRC can be pretty fine.

For the last to uses, it's unfortunate that they are usually implemented as a copylib library. Not as a standalone, dynamic library. It's similar to Base64.

Reply:
I believe that this part can be used like it is, it could be modify in future because the hash is not something that cannot be changed. I do not consider it as a blocking problem.

Reply:
Yesterday I learnt that zlib has crc32() and crc32_z() in its public API. From zlib.h:

~~~~
     uLong crc = crc32(0L, Z_NULL, 0);

     while (read_buffer(buffer, length) != EOF) {
       crc = crc32(crc, buffer, length);
     }
     if (crc != original_crc) error();
~~~~